src/QirRuntime/lib/Tracer/README.md

```diff
@@ -0,0 +1,177 @@
+# Resource Tracer Design Document #
+
+The purpose of the Resource Tracer is to provide efficient and flexible way to estimate resources of a quantum program
+ in QIR representation. The estimates are calculated by simulating execution of the program (as opposed to the static
+ analysis).
+
+In addition to the standard QIR runtime functions, the quantum program will have to:
+
+1. convert _all_ used intrinsic operations into one of the _qis_ operations supported by the tracer (see the list below);
+1. (_optional_) provide callbacks for handling of conditional branches on a measurement (if not provided, the estimates
+ would cover only one branch of the execution);
+1. (_optional_) provide callbacks for start/end of quantum operations (if not provided, all operations will be treated
+ as inlined as if the whole program consisted of a single operation);
+1. (_optional_) provide callbacks for global barriers;
+1. (_optional_) provide description of mapping for frame tracking;
+1. (_optional_) provide names of operations for user friendly output (in the form of `tracer-config.hpp|cpp` files that,
+ ideally, can be generated by the Q# compiler from the mappings described in target.qs).
+
+The last provisions  
+
+The Resource Tracer will consist of:
+
+1. the bridge for the `__quantum__qis__*` extension methods;
+2. the native implementation to back the extensions;
+3. the logic for partitioning gates into layers;
+4. the logic for frame tracking;
+5. output of the collected statistics;
+6. (_lower priority_) the scheduling component to optimize depth and/or width of the circuit.
+
+## List of `__quantum__qis__*` methods ##
+
+___WIP___
+
+| Signature                                             | Description                                                  |
+| :---------------------------------------------------- | :----------------------------------------------------------- |
+| `void __quantum__qis__inject_global_barrier(i32 %id, i32 %duration)` | Function to insert a global barrier. It will be inserted into QIR based on a user defined intrinsic. See [Layering](#layering) section for details. |
+| `void __quantum__qis__on_operation_start(i64 %id)`    | Function to identify the start of a quantum module. The argument is a unique _id_ of the module. The tracer will have an option to treat module boundaries as barriers between layers and (_lower priority_) and option to cache estimates for a module, executed multiple times. The call to the function will be inserted into QIR by the Q# compiler when Tracer is specified as the compilation target. |
+| `void __quantum__qis__on_operation_end(i64 %id)`      | Function to identify the end of a quantum module. The argument is a unique _id_ of the module and must match the _id_ supplied on start of the module. The call to the function will be inserted into QIR by the Q# compiler when Tracer is specified as the compilation target. |

```

Comment:
> void __quantum__qis__on_operation_end(**i64 %id**)

If the modules strictly follow each other or are strictly nested to each other, then the parameter **i64 %id** may be redundant (and can cause questions "_what to do if the parameter does not correspond to the expected one?_").

Reply:
Yes, the modules must be nested, but it should be trivial for the compiler to inject the id into both methods and it will give us an additional invariant to check. I think it's OK to `fail` if the ids don't match.

I don't have strong feelings about this one, though. Could also remove the id from the module_end.

Reply:
Up to you...